File: promo/bridge-intro/render.py
"""index.html を画面なしの Chrome で 1 コマずつ描かせ、そのまま ffmpeg へ流して MP4 にする。

使い方: python3 render.py [--frames 0-749] [--out video.mp4]
音を重ねるのは make_audio.py のあと(最後に mux する)。
"""
import argparse
import base64
import functools
import http.server
import os
import subprocess
import sys
import threading
import time

import imageio_ffmpeg
from playwright.sync_api import sync_playwright

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(HERE, "..", ".."))  # 実画面は images/products/thumbs/ から読むので、リポジトリの根から配る
CHROME = os.environ.get("CHROME", "/opt/pw-browsers/chromium-1194/chrome-linux/chrome")
FFMPEG = imageio_ffmpeg.get_ffmpeg_exe()


def serve():
    class Quiet(http.server.SimpleHTTPRequestHandler):
        def log_message(self, *a):
            pass

    handler = functools.partial(Quiet, directory=REPO)
    httpd = http.server.ThreadingHTTPServer(("127.0.0.1", 0), handler)
    threading.Thread(target=httpd.serve_forever, daemon=True).start()
    return httpd


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--frames", default=None, help="例: 0-99 / 120")
    ap.add_argument("--out", default=os.path.join(HERE, "video.mp4"))
    ap.add_argument("--png-dir", default=None, help="指定すると MP4 ではなく PNG を書き出す(確認用)")
    args = ap.parse_args()

    httpd = serve()
    url = f"http://127.0.0.1:{httpd.server_address[1]}/promo/bridge-intro/index.html?render"

    with sync_playwright() as p:
        browser = p.chromium.launch(executable_path=CHROME)
        page = browser.new_page(viewport={"width": 1920, "height": 1080})
        page.goto(url)
        page.evaluate("window.ready")
        total = page.evaluate("window.TOTAL_FRAMES")
        fps = page.evaluate("TL.fps")

        if args.frames:
            a, _, b = args.frames.partition("-")
            frames = range(int(a), int(b or a) + 1)
        else:
            frames = range(total)

        ff = None
        if args.png_dir:
            os.makedirs(args.png_dir, exist_ok=True)
        else:
            ff = subprocess.Popen(
                [FFMPEG, "-loglevel", "error", "-y", "-f", "image2pipe", "-framerate", str(fps),
                 "-i", "-", "-c:v", "libx264", "-preset", "slow", "-crf", "16",
                 "-pix_fmt", "yuv420p", "-movflags", "+faststart", args.out],
                stdin=subprocess.PIPE,
            )

        t0 = time.time()
        for i in frames:
            data = page.evaluate(f"window.renderFrame({i})")
            png = base64.b64decode(data.split(",", 1)[1])
            if ff:
                ff.stdin.write(png)
            else:
                with open(os.path.join(args.png_dir, f"f{i:04d}.png"), "wb") as fh:
                    fh.write(png)
            if i % 50 == 0:
                print(f"frame {i}/{total}  {time.time() - t0:.0f}s", file=sys.stderr, flush=True)
        if ff:
            ff.stdin.close()
            ff.wait()
        browser.close()
    httpd.shutdown()
    print(f"done in {time.time() - t0:.0f}s", file=sys.stderr)


if __name__ == "__main__":
    main()

File: showreel/render.mjs
// ================================================================
// showreel render — reel.html を 1 フレームずつ描いて MP4 にする
// usage: node showreel/render.mjs [--from 0] [--to 15] [--fps 60] [--samples 6] [--out showreel/showreel.mp4]
//        node showreel/render.mjs --stills 0.5,2.1,7.3 --outdir <dir>   (確認用の静止画だけ)
// 依存: playwright(chromium)・ffmpeg(PATH か FFMPEG)。音は showreel/sound.mjs が書く WAV を重ねる
// フォント: reel.html は Google Fonts を読む。描画時はその要求を curl で取ったキャッシュから返す
//           (Chromium に TLS の例外を与えないため。キャッシュは --fontcache、既定は OS の一時領域)
// ================================================================
import { chromium } from 'playwright';
import { createServer } from 'node:http';
import { readFileSync, existsSync, mkdirSync, writeFileSync } from 'node:fs';
import { execFileSync, spawn } from 'node:child_process';
import { createHash } from 'node:crypto';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, '..');
const argv = Object.fromEntries(process.argv.slice(2).map((a, i, arr) => a.startsWith('--') ? [a.slice(2), arr[i + 1]] : null).filter(Boolean));
const FPS = +(argv.fps || 60);
const FROM = +(argv.from || 0);
const TO = argv.to != null ? +argv.to : null;
const SAMPLES = +(argv.samples || 6);
const SHUTTER = +(argv.shutter || 0.5);            // 1 フレームのうちシャッターが開いている割合(180°)
const OUT = argv.out || path.join(HERE, 'showreel.mp4');
const FFMPEG = process.env.FFMPEG || 'ffmpeg';
const FONTCACHE = argv.fontcache || path.join(os.tmpdir(), 'bridge-showreel-fonts');
const CRF = argv.crf || '14';
mkdirSync(FONTCACHE, { recursive: true });

/* ---- 静的サーバー(canvas に画像を描くため file:// ではなく http で開く) ---- */
const TYPES = { '.html': 'text/html', '.js': 'text/javascript', '.mjs': 'text/javascript', '.png': 'image/png', '.webp': 'image/webp', '.jpg': 'image/jpeg', '.wav': 'audio/wav', '.css': 'text/css', '.svg': 'image/svg+xml' };
const server = createServer((req, res) => {
  const p = path.join(ROOT, decodeURIComponent(req.url.split('?')[0]));
  if (!p.startsWith(ROOT) || !existsSync(p)) { res.writeHead(404); return res.end(); }
  res.writeHead(200, { 'content-type': TYPES[path.extname(p)] || 'application/octet-stream' });
  res.end(readFileSync(p));
});
await new Promise(r => server.listen(0, '127.0.0.1', r));
const BASE = `http://127.0.0.1:${server.address().port}`;

/* ---- フォントの要求は curl のキャッシュから返す ---- */
function cached(url) {
  const f = path.join(FONTCACHE, createHash('sha1').update(url).digest('hex'));
  if (!existsSync(f)) execFileSync('curl', ['-sSfL', '-A', 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/141.0 Safari/537.36', '-o', f, url]);
  return readFileSync(f);
}

const launch = {};
if (process.env.PW_CHROMIUM) launch.executablePath = process.env.PW_CHROMIUM;
const browser = await chromium.launch(launch);
const page = await browser.newPage({ viewport: { width: 1920, height: 1080 }, deviceScaleFactor: 1 });
await page.route(/^https:\/\/fonts\.(googleapis|gstatic)\.com\//, async route => {
  const url = route.request().url();
  try {
    const body = cached(url);
    const type = url.includes('googleapis') ? 'text/css' : 'font/woff2';
    await route.fulfill({ status: 200, body, headers: { 'content-type': type, 'access-control-allow-origin': '*' } });
  } catch (e) { console.error('font fetch failed', url); await route.abort(); }
});
page.on('console', m => { if (m.type() === 'error' || m.type() === 'warning') console.log('[page]', m.text()); });
page.on('pageerror', e => console.log('[pageerror]', e.message));

await page.goto(`${BASE}/showreel/reel.html?render=1`, { waitUntil: 'load' });
await page.evaluate(() => window.REEL.ready);
const DUR = await page.evaluate(() => window.REEL.DUR);

/* ---- 1 フレーム: サブフレームを Float で平均して動きのぼけを作る ---- */
async function frame(t, samples) {
  return page.evaluate(async ([t, samples, shutter, fps]) => {
    const R = window.REEL;
    R.renderBlur(t, samples, shutter / fps);
    const blob = await new Promise(r => R.out.toBlob(r, 'image/png'));
    const buf = new Uint8Array(await blob.arrayBuffer());
    let s = ''; for (let i = 0; i < buf.length; i += 0x8000) s += String.fromCharCode.apply(null, buf.subarray(i, i + 0x8000));
    return btoa(s);
  }, [t, samples, SHUTTER, FPS]);
}

if (argv.stills) {
  const dir = argv.outdir || path.join(os.tmpdir(), 'showreel-stills');
  mkdirSync(dir, { recursive: true });
  for (const ts of argv.stills.split(',').map(Number)) {
    const b64 = await frame(ts, +(argv.samples || 1));
    const f = path.join(dir, `t${ts.toFixed(3).padStart(6, '0')}.png`);
    writeFileSync(f, Buffer.from(b64, 'base64'));
    console.log(f);
  }
  await browser.close(); server.close();
  process.exit(0);
}

/* ---- 動画: PNG を ffmpeg に流す。音(sound.wav)があれば重ねる ---- */
const end = TO ?? DUR;
const n0 = Math.round(FROM * FPS), n1 = Math.round(end * FPS);
const wav = path.join(HERE, 'sound.wav');
const withAudio = existsSync(wav) && !argv.noaudio;
const args = ['-y', '-loglevel', 'error', '-f', 'image2pipe', '-framerate', String(FPS), '-c:v', 'png', '-i', '-'];
if (withAudio) args.push('-ss', String(FROM), '-t', String((n1 - n0) / FPS), '-i', wav);
// RGB→YUV は bt709 の行列で(既定の bt601 のままだと藍・砂がわずかにずれる)
args.push('-vf', 'scale=out_color_matrix=bt709:out_range=tv:flags=lanczos+accurate_rnd+full_chroma_int',
  '-c:v', 'libx264', '-preset', 'slow', '-crf', CRF, '-pix_fmt', 'yuv420p', '-profile:v', 'high', '-tune', 'animation',
  '-colorspace', 'bt709', '-color_primaries', 'bt709', '-color_trc', 'bt709', '-movflags', '+faststart');
if (withAudio) args.push('-c:a', 'aac', '-b:a', '256k', '-shortest');
args.push(OUT);
const ff = spawn(FFMPEG, args, { stdio: ['pipe', 'inherit', 'inherit'] });
const t0 = Date.now();
for (let n = n0; n < n1; n++) {
  const b64 = await frame(n / FPS, SAMPLES);
  if (!ff.stdin.write(Buffer.from(b64, 'base64'))) await new Promise(r => ff.stdin.once('drain', r));
  if (n % 60 === 0) console.log(`frame ${n}/${n1} (${((Date.now() - t0) / 1000).toFixed(0)}s)`);
}
ff.stdin.end();
await new Promise(r => ff.on('close', r));
await browser.close(); server.close();
console.log('wrote', OUT);
